### PR TITLE
Configapi uma protect

### PR DESCRIPTION
**Pull request by pujavs**
@yuriyz ,
I have refactored code, tested the oauth2 flow and cleaned up the code for as well as improved  logging.
Request you to please review.

### CODE REVIEW

File: src/main/java/io/jans/configapi/filters/AuthorizationFilter.java [hunk 1]
```diff
@@ -47,33 +48,52 @@
     @Context
     private ResourceInfo resourceInfo;
     
-
     @Inject
     AuthorizationService authorizationService;
+    
+    @Inject
+    ApiTestMode apiTestMode;
```
Comment (yuriyz) [inline src/main/java/io/jans/configapi/filters/AuthorizationFilter.java:55]: We can't use test code in production. Test code should live within tests only. Refactor it to traditional setUp/tearDown.
For this particular case it should be on suite level. So we pick up what we need one time before whole suite and then tear it down when it finishes.

File: src/main/java/io/jans/configapi/filters/AuthorizationFilter.java [hunk 2]
```diff
@@ -47,33 +48,52 @@
...
+            log.info("\n\n\n AuthorizationFilter::filter() - after testAuthenticationPrep() -  authorizationHeader = " + authorizationHeader+" , apiTestMode.isTestMode() = "+apiTestMode.isTestMode()+"\n\n\n");
+            
+            //Api protection validation
             this.authorizationService.processAuthorization(authorizationHeader, resourceInfo, context.getMethod(), request.getRequestURI());
             log.info("======AUTHORIZATION  GRANTED===========================================");
         } catch (Exception ex) {
             log.error("======AUTHORIZATION  FAILED ===========================================", ex);
             abortWithUnauthorized(context);
         }
+        finally {
```
Comment (yuriyz) [inline src/main/java/io/jans/configapi/filters/AuthorizationFilter.java:88]: It must be in tearDown of the suite. In junit.jupiter I guess it's `@AfterAll`, no ?

File: src/main/java/io/jans/configapi/filters/AuthorizationFilter.java [hunk 3]
```diff
@@ -86,5 +106,11 @@ private void abortWithUnauthorized(ContainerRequestContext requestContext) {
         requestContext.abortWith(Response.status(Response.Status.UNAUTHORIZED)
                 .header(HttpHeaders.WWW_AUTHENTICATE, AUTHENTICATION_SCHEME).build());
     }
+    
+    private String testAuthenticationPrep(ResourceInfo resourceInfo, String method, String path) throws Exception {
```
Comment (yuriyz) [inline src/main/java/io/jans/configapi/filters/AuthorizationFilter.java:110]: Must be in test sources as well as ApiTestMode.

File: src/main/java/io/jans/configapi/auth/UmaResourceProtectionService.java [hunk 4]
```diff
@@ -112,26 +105,31 @@ private void createScopeIfNeeded() {
                         }
                     }
 
+                    ScopeType scopeType = ScopeType.UMA;
+                    /*
+                     * if(ApiConstants.PROTECTION_TYPE_OAUTH2.equals(ConfigurationFactory.
+                     * getApiProtectionType())) { scopeType = ScopeType.OPENID; }
+                     */
```
Comment (yuriyz) [inline src/main/java/io/jans/configapi/auth/UmaResourceProtectionService.java:112]: Remove commented code.

File: src/main/java/io/jans/configapi/auth/client/AuthClientFactory.java [hunk 5]
```diff
@@ -61,6 +71,174 @@ public static IntrospectionResponse getIntrospectionResponse(String url, String
...
+                .register(engine);
+        restClient.property("Authorization", "Basic " + tokenRequest.getEncodedCredentials());
+        restClient.property("Content-Type", MediaType.APPLICATION_FORM_URLENCODED);
+
+        ResteasyWebTarget target = (ResteasyWebTarget) ResteasyClientBuilder.newClient(restClient.getConfiguration())
+                .target(revokeTokenUrl);
+
+        
+       // Response response = request
+         //       .post(Entity.form(new MultivaluedHashMap<String, String>(tokenRequest.getParameters())));
```
Comment (yuriyz) [inline src/main/java/io/jans/configapi/auth/client/AuthClientFactory.java:100]: Remove redundant commented lines.

File: src/main/java/io/jans/configapi/auth/client/AuthClientFactory.java [hunk 6]
```diff
@@ -61,6 +71,174 @@ public static IntrospectionResponse getIntrospectionResponse(String url, String
...
+                .register(engine);
+        restClient.property("Authorization", "Basic " + tokenRequest.getEncodedCredentials());
+        restClient.property("Content-Type", MediaType.APPLICATION_FORM_URLENCODED);
+
+        ResteasyWebTarget target = (ResteasyWebTarget) ResteasyClientBuilder.newClient(restClient.getConfiguration())
+                .target(tokenUrl);
+
+        
+       // Response response = request
+         //       .post(Entity.form(new MultivaluedHashMap<String, String>(tokenRequest.getParameters())));
```
Comment (yuriyz) [inline src/main/java/io/jans/configapi/auth/client/AuthClientFactory.java:144]: Redundant lines.

File: src/main/java/io/jans/configapi/auth/service/UmaService.java [hunk 7]
```diff
@@ -86,33 +86,36 @@ public UmaRptIntrospectionService getUmaRptIntrospectionService() {
     }
 
     public void validateRptToken(Token patToken, String authorization, String resourceId, List<String> scopeIds) {
-        log.trace("Validating RPT, patToken:{}, authorization:{}, resourceId: {}, scopeIds: {} ", patToken, authorization, resourceId, scopeIds);
+        log.trace("Validating RPT, patToken: {}, authorization: {}, resourceId: {}, scopeIds: {} ", patToken,
+                authorization, resourceId, scopeIds);
+        
         if (patToken == null) {
-            log.info("Token is blank"); 
+            log.warn("Token is blank");
```
Comment (yuriyz) [inline src/main/java/io/jans/configapi/auth/service/UmaService.java:93]: trace

File: src/main/java/io/jans/configapi/auth/util/ApiTestMode.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+import javax.inject.Inject;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.container.ResourceInfo;
+import javax.ws.rs.core.Response;
+
+@ApplicationScoped
+public class ApiTestMode {
```
Comment (yuriyz) [inline src/main/java/io/jans/configapi/auth/util/ApiTestMode.java:66]: Test code must be in test sources.

**Review by yuriyz — CHANGES_REQUESTED** [review]
We shouldn't put test code in application source code. It has to be in test sources. Also use suite setup and tear down.

File: src/main/java/io/jans/configapi/filters/AuthorizationFilter.java (hunk 1 above)
Comment (pujavs) [reply src/main/java/io/jans/configapi/filters/AuthorizationFilter.java]: Modified the code to move the test code within test folder.
jans-config-api uses Quarkus and Karate framework making it complex and no proper hooks to implement.
If only Quarkus was used and also test cases were written as QaurkusTest  we could have used the QuarkusTestCallbacks like QuarkusTestBeforeAllCallback, QuarkusTestBeforeEachCallback,QuarkusTestAfterEachCallback OR Junit Test then we could have used the setup/tearDown with either Junit @BeforeAll, @AfterAll, @BeforeEach.
But at present we have one single Test class KarateTestRunner/JenkinsTestRunner and this is called only once thus token generation for each feature test scenario method is not possible.
So after much exploration i could come up with Karate hooks
**Our requirement:** 
1) We need test client 
2) Appropriate token for each scenario method that
**Implementations:**
1) Before execution of Karate feature create client :- During Karate config setup the client is created. This is implemented in karate-config.js for local and karate-config-jenkins.js using testClient.feature for jenkins. Status local is tested, jenkins that is via karate-config-jenkins.js is pending
2) accessToken generation and passing in header :- Required before invoking each endpoint method. Code in each feature file calling token.feature to generate accessToken. Status tested locally for three features attributes.feature, cache.feature and clients.feature. Jenkins env is pending. 
Once you approve the change plan is to enable the same for all features.

File: src/main/java/io/jans/configapi/filters/AuthorizationFilter.java (hunk 2 above)
Comment (pujavs) [reply src/main/java/io/jans/configapi/filters/AuthorizationFilter.java]: Removed the same.

File: src/main/java/io/jans/configapi/filters/AuthorizationFilter.java (hunk 3 above)
Comment (pujavs) [reply src/main/java/io/jans/configapi/filters/AuthorizationFilter.java]: Moved the test code from main java code AuthorizationFilter class to test src code.

File: src/main/java/io/jans/configapi/auth/UmaResourceProtectionService.java (hunk 4 above)
Comment (pujavs) [reply src/main/java/io/jans/configapi/auth/UmaResourceProtectionService.java]: Removed commented code

File: src/main/java/io/jans/configapi/auth/client/AuthClientFactory.java (hunk 5 above)
Comment (pujavs) [reply src/main/java/io/jans/configapi/auth/client/AuthClientFactory.java]: Removed commented code

File: src/main/java/io/jans/configapi/auth/client/AuthClientFactory.java (hunk 6 above)
Comment (pujavs) [reply src/main/java/io/jans/configapi/auth/client/AuthClientFactory.java]: Removed the redundant code

File: src/main/java/io/jans/configapi/auth/service/UmaService.java (hunk 7 above)
Comment (pujavs) [reply src/main/java/io/jans/configapi/auth/service/UmaService.java]: changed to trace

File: src/main/java/io/jans/configapi/auth/util/ApiTestMode.java (hunk 8 above)
Comment (pujavs) [reply src/main/java/io/jans/configapi/auth/util/ApiTestMode.java]: Agreed, have moved the test code to test src

**Comment by pujavs** [comment]
@yuriyz , request you to please review and share feedback. If you are find with the changes then request approve of merge request

**Review by yuriyz — APPROVED** [review]

**Pull request merged**